### DISCUSSION TITLE
which function sends information to the aggregation function?

### DISCUSSION THREAD

**Discussion by dskhanirfan** [securefederatedai/openfederatedlearning#346] [General]
what is the name of the function that is run on the participating collaborators which sends information to the aggregation function? Can you outline the function calls in a sequential manner i-e one function call after the other that leads to final aggregation in an FL round?

**Comment by ayyaz-azeem** [comment]
which function dispatches parameters run on individual collaborators to the aggregator function? @tanertopal @argretzi @kshannon

**Comment by itrushkin** [comment]
Hi @dskhanirfan! Thank you for your interest in OpenFL!

We perform aggregation at the end of the round. After all collaborators reported their last task results to aggregator, we start aggregation.
Sequence:
1. [`Collaborator.do_task()`](https://github.com/intel/openfl/blob/develop/openfl/component/collaborator/collaborator.py#L184) - Local task execution.
2. [`Collaborator.send_task_results()`](https://github.com/intel/openfl/blob/develop/openfl/component/collaborator/collaborator.py#L392) - Send the results to collaborator. Probably this is the method you are looking for.
3. [`AggregatorGRPCClient.send_local_task_results()`](https://github.com/intel/openfl/blob/fea20b9b983bc95a4c6d334e456db2114b084bef/openfl/transport/grpc/aggregator_client.py#L290) - Wrap the data into gRPC request.
4. [`AggregatorGRPCServer.SendLocalTaskResults()`](https://github.com/intel/openfl/blob/fea20b9b983bc95a4c6d334e456db2114b084bef/openfl/transport/grpc/aggregator_server.py#L199) - Unwrap the request.
5. [`Aggregator.send_local_task_results()`](https://github.com/intel/openfl/blob/fea20b9b983bc95a4c6d334e456db2114b084bef/openfl/component/aggregator/aggregator.py#L461) - Process task results.
6. [`Aggregator._end_of_task_check()`](https://github.com/intel/openfl/blob/fea20b9b983bc95a4c6d334e456db2114b084bef/openfl/component/aggregator/aggregator.py#L634) - Check whether task is completed on all collaborators.
7. [`Aggregator._end_of_round_check()`](https://github.com/intel/openfl/blob/fea20b9b983bc95a4c6d334e456db2114b084bef/openfl/component/aggregator/aggregator.py#L828) - Check whether the task was the last task in round.
8. [`Aggregator._compute_validation_related_task_metrics()`](https://github.com/intel/openfl/blob/fea20b9b983bc95a4c6d334e456db2114b084bef/openfl/component/aggregator/aggregator.py#L746) - Preparing the data for aggregation.
9. [`TensorDB.get_aggregated_tensor()`](https://github.com/intel/openfl/blob/fea20b9b983bc95a4c6d334e456db2114b084bef/openfl/databases/tensor_db.py#L106) - The aggregation itself. On line 171 `aggregation_function` is called.

**Reply by dskhanirfan** [reply]
Thank you @itrushkin can you kindly mention the filenames which contain the functions?

**Reply by itrushkin** [reply]
You can click on each method in the list above to find its location in code. Locations:

Steps 1-2: openfl/component/collaborator/collaborator.py
3: openfl/transport/grpc/aggregator_client.py
4: openfl/transport/grpc/aggregator_server.py
5-8: openfl/component/aggregator/aggregator.py
9: openfl/databases/tensor_db.py
